src/nomad_simulations/schema_packages/data_types.py

```diff
@@ -0,0 +1,260 @@
+#
+# Copyright The NOMAD Authors.
+#
+# This file is part of NOMAD. See https://nomad-lab.eu for further info.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from abc import ABC, abstractmethod
+from collections.abc import Generator
+from typing import Any, Callable, TypeVar, Union
+
+import numpy as np
+from nomad.metainfo.data_type import ExactNumber, InexactNumber
+
+T = TypeVar('T', bound='BoundedDataType')
+
+
+def _flatten_values(data: Any) -> Generator[Any, None, None]:
+    """Generator that yields all scalar values from nested list/array structure."""
+    if isinstance(data, np.ndarray):
+        yield from data.flatten()
+    elif isinstance(data, list):
+        for item in data:
+            if isinstance(item, list):
+                yield from _flatten_values(item)
+            else:
+                yield item
+    else:
+        yield data
+
+
+def _check_bounds(
+    value: Union[int, float],
+    min_value: float,
+    max_value: float,
+    min_inclusive: bool,
+    max_inclusive: bool,
+) -> bool:
+    """Check if a single value is within the specified bounds."""
+    # lower bound
+    if np.isfinite(min_value):
+        if min_inclusive:
+            if value < min_value:
+                return False
+        else:
+            if value <= min_value:
+                return False
+
+    # upper bound
+    if np.isfinite(max_value):
+        if max_inclusive:
+            if value > max_value:
+                return False
+        else:
+            if value >= max_value:
+                return False
+
+    return True
+
+
+def _get_bounds_str(
+    min_value: float, max_value: float, min_inclusive: bool, max_inclusive: bool
+) -> str:
+    """Get string representation of bounds."""
+    left = '[' if min_inclusive else '('
+    right = ']' if max_inclusive else ')'
+
+    min_str = str(min_value) if np.isfinite(min_value) else '-∞'
+    max_str = str(max_value) if np.isfinite(max_value) else '∞'
+
+    return f'{left}{min_str}, {max_str}{right}'
+
+
+class BoundedDataType(ABC):
```

Comment:
This thing looks unnecessarily over-engineered.
One can directly use for example, `BoundedFloat(dtype=float,range='[0,3)')` to provide all necessary information.
I do not see the benefits of `BoundedFloat().min(0,inclusive=True).max(3,inclusive=False)`.
But I am okay with the current impl.

Reply:
So, atm I handle the bound checking differently for `NaN`, depending on whether `dtype` is a `float` (pass) or `int` (error). Can I delegate this to the NOMAD data types?